File: android/app/src/main/java/uk/unmannedsystems/dpm_android/camera/PropertyLoader.kt
package uk.unmannedsystems.dpm_android.camera

import android.content.Context
import android.util.Log
import org.json.JSONObject

/**
 * PropertyLoader - Loads and validates camera property values from specification JSON
 *
 * This class implements the specification-first architecture by loading camera
 * property values from assets/camera_properties.json at runtime.
 *
 * Background:
 * Previously, property values (ISO, shutter speed, aperture) were hardcoded
 * independently in Air-Side C++ and Ground-Side Android, causing synchronization
 * failures. This class ensures both sides use identical value lists from a
 * single source of truth.
 *
 * Usage:
 *   PropertyLoader.initialize(context)
 *   if (!PropertyLoader.isInitialized()) {
 *       // Handle error
 *   }
 *   val isoValues = PropertyLoader.getIsoValues()
 *   val isValid = PropertyLoader.isValidValue("iso", "800")
 *
 * See: docs/CAMERA_PROPERTIES_FIX_TRACKING.md for complete context
 * See: docs/CC_READ_THIS_FIRST.md lines 29-111 for specification-first rules
 */
object PropertyLoader {
    private const val TAG = "PropertyLoader"
    private const val JSON_FILENAME = "camera_properties.json"

    private var initialized = false
    private val isoValues = mutableSetOf<String>()
    private val shutterSpeedValues = mutableSetOf<String>()
    private val apertureValues = mutableSetOf<String>()

    /**
     * Initialize PropertyLoader by loading camera_properties.json from assets
     *
     * @param context Android context for accessing assets
     * @return true if initialization successful, false on error
     *
     * This should be called once at application startup, before any camera
     * operations. Logs detailed error messages on failure.
     */
    fun initialize(context: Context): Boolean {
        if (initialized) {
            Log.w(TAG, "PropertyLoader.initialize() called multiple times - ignoring")
            return true
        }

        Log.i(TAG, "PropertyLoader: Loading camera properties from $JSON_FILENAME")

        try {
            // Load JSON from assets
            val jsonString = context.assets.open(JSON_FILENAME).bufferedReader().use {
                it.readText()
            }

            val spec = JSONObject(jsonString)

            // Validate JSON structure
            if (!spec.has("properties")) {
                Log.e(TAG, "PropertyLoader: Invalid JSON - missing 'properties' field")
                return false
            }

            val properties = spec.getJSONObject("properties")

            // Load ISO values
            if (properties.has("iso")) {
                val isoProperty = properties.getJSONObject("iso")
                if (isoProperty.has("validation")) {
                    val validation = isoProperty.getJSONObject("validation")
                    if (validation.has("values")) {
                        val values = validation.getJSONArray("values")
                        for (i in 0 until values.length()) {
                            isoValues.add(values.getString(i))
                        }
                        Log.i(TAG, "PropertyLoader: Loaded ${isoValues.size} ISO values")
                    } else {
                        Log.e(TAG, "PropertyLoader: ISO validation missing 'values'")
                        return false
                    }
                } else {
                    Log.e(TAG, "PropertyLoader: ISO property missing 'validation'")
                    return false
                }
            } else {
                Log.e(TAG, "PropertyLoader: Missing 'iso' property in JSON")
                return false
            }

            // Load Shutter Speed values
            if (properties.has("shutter_speed")) {
                val shutterProperty = properties.getJSONObject("shutter_speed")
                if (shutterProperty.has("validation")) {
                    val validation = shutterProperty.getJSONObject("validation")
                    if (validation.has("values")) {
                        val values = validation.getJSONArray("values")
                        for (i in 0 until values.length()) {
                            shutterSpeedValues.add(values.getString(i))
                        }
                        Log.i(TAG, "PropertyLoader: Loaded ${shutterSpeedValues.size} shutter speed values")
                    } else {
                        Log.e(TAG, "PropertyLoader: Shutter speed validation missing 'values'")
                        return false
                    }
                } else {
                    Log.e(TAG, "PropertyLoader: Shutter speed property missing 'validation'")
                    return false
                }
            } else {
                Log.e(TAG, "PropertyLoader: Missing 'shutter_speed' property in JSON")
                return false
            }

            // Load Aperture values
            if (properties.has("aperture")) {
                val apertureProperty = properties.getJSONObject("aperture")
                if (apertureProperty.has("validation")) {
                    val validation = apertureProperty.getJSONObject("validation")
                    if (validation.has("values")) {
                        val values = validation.getJSONArray("values")
                        for (i in 0 until values.length()) {
                            apertureValues.add(values.getString(i))
                        }
                        Log.i(TAG, "PropertyLoader: Loaded ${apertureValues.size} aperture values")
                    } else {
                        Log.e(TAG, "PropertyLoader: Aperture validation missing 'values'")
                        return false
                    }
                } else {
                    Log.e(TAG, "PropertyLoader: Aperture property missing 'validation'")
                    return false
                }
            } else {
                Log.e(TAG, "PropertyLoader: Missing 'aperture' property in JSON")
                return false
            }

            // Validation: Ensure we loaded expected counts
            if (isoValues.size < 10) {
                Log.w(TAG, "PropertyLoader: Only loaded ${isoValues.size} ISO values - expected ~35")
            }
            if (shutterSpeedValues.size < 10) {
                Log.w(TAG, "PropertyLoader: Only loaded ${shutterSpeedValues.size} shutter speed values - expected ~56")
            }
            if (apertureValues.size < 5) {
                Log.w(TAG, "PropertyLoader: Only loaded ${apertureValues.size} aperture values - expected ~23")
            }

            initialized = true
            Log.i(TAG, "PropertyLoader: Initialization complete")
            Log.i(TAG, "PropertyLoader: Loaded total of ${isoValues.size + shutterSpeedValues.size + apertureValues.size} property values from specification")

            return true

        } catch (e: Exception) {
            Log.e(TAG, "PropertyLoader: Unexpected error during initialization", e)
            return false
        }
    }

    /**
     * Check if PropertyLoader has been successfully initialized
     *
     * @return true if initialize() completed successfully
     */
    fun isInitialized(): Boolean {
        return initialized
    }

    /**
     * Get set of valid ISO values from specification
     *
     * @return Set of valid ISO value strings (e.g., "auto", "100", "200", ...)
     *
     * Returns all 35 ISO values including:
     * - "auto" (0xFFFFFFFF)
     * - Extended low: "50", "64", "80"
     * - Standard range: "100" through "102400" (full and third stops)
     */
    fun getIsoValues(): Set<String> {
        if (!initialized) {
            Log.e(TAG, "PropertyLoader.getIsoValues() called before initialization!")
            return emptySet()
        }
        return isoValues.toSet()
    }

    /**
     * Get set of valid shutter speed values from specification
     *
     * @return Set of valid shutter speed strings (e.g., "1/8000", "0.3\"", ...)
     *
     * Returns all 56 shutter speed values:
     * - 35 fast speeds: "1/8000" through "1/3"
     * - 21 long exposures: "0.3\"" through "30\""
     *
     * Note: AUTO and BULB modes are disabled for UAV safety
     */
    fun getShutterSpeedValues(): Set<String> {
        if (!initialized) {
            Log.e(TAG, "PropertyLoader.getShutterSpeedValues() called before initialization!")
            return emptySet()
        }
        return shutterSpeedValues.toSet()
    }

    /**
     * Get set of valid aperture (f-stop) values from specification
     *
     * @return Set of valid aperture strings (e.g., "f/1.4", "f/2.8", ...)
     *
     * Returns all 23 aperture values from f/1.4 to f/22
     */
    fun getApertureValues(): Set<String> {
        if (!initialized) {
            Log.e(TAG, "PropertyLoader.getApertureValues() called before initialization!")
            return emptySet()
        }
        return apertureValues.toSet()
    }

    /**
     * Validate that a property value exists in specification
     *
     * @param property Property name ("iso", "shutter_speed", "aperture")
     * @param value Value string to validate
     * @return true if value is valid for this property
     */
    fun isValidValue(property: String, value: String): Boolean {
        if (!initialized) {
            Log.e(TAG, "PropertyLoader.isValidValue() called before initialization!")
            return false
        }

        return when (property) {
            "iso" -> isoValues.contains(value)
            "shutter_speed" -> shutterSpeedValues.contains(value)
            "aperture" -> {
                // Handle both "f/2.8" and "2.8" formats
                val normalizedValue = value.removePrefix("f/")
                apertureValues.any { it.removePrefix("f/") == normalizedValue }
            }
            else -> {
                Log.w(TAG, "PropertyLoader.isValidValue() called with unknown property: $property")
                false
            }
        }
    }

    /**
     * Get count of valid values for a property (for diagnostics)
     *
     * @param property Property name ("iso", "shutter_speed", "aperture")
     * @return Number of valid values, or 0 if property unknown
     */
    fun getValueCount(property: String): Int {
        if (!initialized) {
            return 0
        }

        return when (property) {
            "iso" -> isoValues.size
            "shutter_speed" -> shutterSpeedValues.size
            "aperture" -> apertureValues.size
            else -> 0
        }
    }
}

File: sbc/src/camera/property_loader.cpp
#include "property_loader.h"
#include "../utils/logger.h"
#include <nlohmann/json.hpp>
#include <fstream>
#include <sstream>
#include "logging/structured_logger.h"

using json = nlohmann::json;

// Initialize static members
bool PropertyLoader::initialized_ = false;
std::unordered_set<std::string> PropertyLoader::iso_values_;
std::unordered_set<std::string> PropertyLoader::shutter_speed_values_;
std::unordered_set<std::string> PropertyLoader::aperture_values_;

bool PropertyLoader::initialize(const std::string& json_path) {
    if (initialized_) {
        LOG_WARNING(LogContext::CAMERA, "PropertyLoader::initialize() called multiple times - ignoring");
        return true;
    }

    LOG_INFO(LogContext::CAMERA, "PropertyLoader: Loading camera properties from " + json_path);

    try {
        // Load JSON file
        std::ifstream file(json_path);
        if (!file.is_open()) {
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Failed to open " + json_path);
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Make sure camera_properties.json exists at ~/DPM-V2/protocol/");
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: This is a shared specification file - NEVER in docs/ folder!");
            return false;
        }

        json spec;
        file >> spec;
        file.close();

        // Validate JSON structure
        if (!spec.contains("properties")) {
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Invalid JSON - missing 'properties' field");
            return false;
        }

        json properties = spec["properties"];

        // Load ISO values
        if (properties.contains("iso")) {
            auto iso_property = properties["iso"];
            if (iso_property.contains("validation") && iso_property["validation"].contains("values")) {
                auto values = iso_property["validation"]["values"];
                if (values.is_array()) {
                    for (const auto& val : values) {
                        if (val.is_string()) {
                            iso_values_.insert(val.get<std::string>());
                        }
                    }
                    LOG_INFO(LogContext::CAMERA, "PropertyLoader: Loaded " + std::to_string(iso_values_.size()) + " ISO values");
                } else {
                    LOG_ERROR(LogContext::CAMERA, "PropertyLoader: ISO validation.values is not an array");
                    return false;
                }
            } else {
                LOG_ERROR(LogContext::CAMERA, "PropertyLoader: ISO property missing validation.values");
                return false;
            }
        } else {
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Missing 'iso' property in JSON");
            return false;
        }

        // Load Shutter Speed values
        if (properties.contains("shutter_speed")) {
            auto shutter_property = properties["shutter_speed"];
            if (shutter_property.contains("validation") && shutter_property["validation"].contains("values")) {
                auto values = shutter_property["validation"]["values"];
                if (values.is_array()) {
                    for (const auto& val : values) {
                        if (val.is_string()) {
                            shutter_speed_values_.insert(val.get<std::string>());
                        }
                    }
                    LOG_INFO(LogContext::CAMERA, "PropertyLoader: Loaded " + std::to_string(shutter_speed_values_.size()) + " shutter speed values");
                } else {
                    LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Shutter speed validation.values is not an array");
                    return false;
                }
            } else {
                LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Shutter speed property missing validation.values");
                return false;
            }
        } else {
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Missing 'shutter_speed' property in JSON");
            return false;
        }

        // Load Aperture values
        if (properties.contains("aperture")) {
            auto aperture_property = properties["aperture"];
            if (aperture_property.contains("validation") && aperture_property["validation"].contains("values")) {
                auto values = aperture_property["validation"]["values"];
                if (values.is_array()) {
                    for (const auto& val : values) {
                        if (val.is_string()) {
                            aperture_values_.insert(val.get<std::string>());
                        }
                    }
                    LOG_INFO(LogContext::CAMERA, "PropertyLoader: Loaded " + std::to_string(aperture_values_.size()) + " aperture values");
                } else {
                    LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Aperture validation.values is not an array");
                    return false;
                }
            } else {
                LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Aperture property missing validation.values");
                return false;
            }
        } else {
            LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Missing 'aperture' property in JSON");
            return false;
        }

        // Validation: Ensure we loaded expected counts
        if (iso_values_.size() < 10) {
            LOG_WARNING(LogContext::CAMERA, "PropertyLoader: Only loaded " + std::to_string(iso_values_.size()) + " ISO values - expected ~35");
        }
        if (shutter_speed_values_.size() < 10) {
            LOG_WARNING(LogContext::CAMERA, "PropertyLoader: Only loaded " + std::to_string(shutter_speed_values_.size()) + " shutter speed values - expected ~56");
        }
        if (aperture_values_.size() < 5) {
            LOG_WARNING(LogContext::CAMERA, "PropertyLoader: Only loaded " + std::to_string(aperture_values_.size()) + " aperture values - expected ~23");
        }

        initialized_ = true;
        LOG_INFO(LogContext::CAMERA, "PropertyLoader: Initialization complete");
        LOG_INFO(LogContext::CAMERA, "PropertyLoader: Loaded total of " +
                    std::to_string(iso_values_.size() + shutter_speed_values_.size() + aperture_values_.size()) +
                    " property values from specification");

        return true;

    } catch (const json::parse_error& e) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader: JSON parse error: " + std::string(e.what()));
        return false;
    } catch (const json::type_error& e) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader: JSON type error: " + std::string(e.what()));
        return false;
    } catch (const std::exception& e) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader: Unexpected error: " + std::string(e.what()));
        return false;
    }
}

bool PropertyLoader::isInitialized() {
    return initialized_;
}

const std::unordered_set<std::string>& PropertyLoader::getIsoValues() {
    if (!initialized_) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader::getIsoValues() called before initialization!");
        static std::unordered_set<std::string> empty;
        return empty;
    }
    return iso_values_;
}

const std::unordered_set<std::string>& PropertyLoader::getShutterSpeedValues() {
    if (!initialized_) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader::getShutterSpeedValues() called before initialization!");
        static std::unordered_set<std::string> empty;
        return empty;
    }
    return shutter_speed_values_;
}

const std::unordered_set<std::string>& PropertyLoader::getApertureValues() {
    if (!initialized_) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader::getApertureValues() called before initialization!");
        static std::unordered_set<std::string> empty;
        return empty;
    }
    return aperture_values_;
}

bool PropertyLoader::isValidValue(const std::string& property, const std::string& value) {
    if (!initialized_) {
        LOG_ERROR(LogContext::CAMERA, "PropertyLoader::isValidValue() called before initialization!");
        return false;
    }

    if (property == "iso") {
        return iso_values_.find(value) != iso_values_.end();
    } else if (property == "shutter_speed") {
        return shutter_speed_values_.find(value) != shutter_speed_values_.end();
    } else if (property == "aperture") {
        return aperture_values_.find(value) != aperture_values_.end();
    } else {
        LOG_WARNING(LogContext::CAMERA, "PropertyLoader::isValidValue() called with unknown property: " + property);
        return false;
    }
}

size_t PropertyLoader::getValueCount(const std::string& property) {
    if (!initialized_) {
        return 0;
    }

    if (property == "iso") {
        return iso_values_.size();
    } else if (property == "shutter_speed") {
        return shutter_speed_values_.size();
    } else if (property == "aperture") {
        return aperture_values_.size();
    } else {
        return 0;
    }
}
